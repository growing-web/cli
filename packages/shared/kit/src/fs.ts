import path from 'pathe'
import fs from 'fs-extra'

export function findup<T>(
  rootDir: string,
  fn: (dir: string) => T | undefined,
): T | null {
  let dir = rootDir
  while (dir !== path.dirname(dir)) {
    const res = fn(dir)
    if (res) {
      return res
    }
    dir = path.dirname(dir)
  }
  return null
}

export function lookupFile(
  dir: string,
  formats: string[],
  pathOnly = false,
): string | undefined {
  for (const format of formats) {
    const fullPath = path.join(dir, format)
    if (fs.existsSync(fullPath) && fs.statSync(fullPath).isFile()) {
      return pathOnly ? fullPath : fs.readFileSync(fullPath, 'utf-8')
    }
  }
  const parentDir = path.dirname(dir)
  if (parentDir !== dir) {
    return lookupFile(parentDir, formats, pathOnly)
  }
}

export function JSONReader(
  filename: string,
  silent: boolean = true,
): Record<string, any> {
  try {
    return fs.readJSONSync(filename, { encoding: 'utf-8' })
  } catch (error: any) {
    if (silent) {
      return {}
    } else {
      throw new Error(error)
    }
  }
}
