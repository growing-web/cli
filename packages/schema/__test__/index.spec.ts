import { test, describe, expect } from 'vitest'
import path from 'path'
import fs from 'fs-extra'

describe('schema test', () => {
  test('index.json is exits.', () => {
    const schema = path.resolve(__dirname, '../index.json')
    const content = fs.readJSONSync(schema)
    expect(content).toMatchSnapshot()
  })
})
