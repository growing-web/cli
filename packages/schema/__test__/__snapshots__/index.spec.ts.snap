// Vitest Snapshot v1

exports[`schema test > index.json is exits. 1`] = `
{
  "$id": "http://growing-web.manifest/schema.json",
  "$schema": "http://json-schema.org/draft-07/schema#",
  "additionalProperties": false,
  "definitions": {
    "banner": {
      "additionalProperties": false,
      "properties": {
        "footer": {
          "type": "string",
        },
        "header": {
          "type": "string",
        },
      },
      "type": "object",
    },
    "entries": {
      "items": {
        "$ref": "#/definitions/entry",
      },
      "minLength": 1,
      "type": "array",
    },
    "entry": {
      "additionalProperties": false,
      "properties": {
        "input": {
          "type": "string",
        },
        "output": {
          "$ref": "#/definitions/output",
        },
        "publicPath": {
          "type": "string",
        },
      },
      "required": [
        "input",
      ],
      "type": "object",
    },
    "formats": {
      "items": {
        "enum": [
          "cjs",
          "esm",
          "iife",
          "umd",
          "system",
        ],
        "type": "string",
      },
      "type": "array",
    },
    "meta": {
      "additionalProperties": false,
      "properties": {
        "umdName": {
          "type": "string",
        },
      },
      "type": "object",
    },
    "output": {
      "additionalProperties": false,
      "properties": {
        "assetFileNames": {
          "type": "string",
        },
        "banner": {
          "$ref": "#/definitions/banner",
        },
        "chunkFileNames": {
          "type": "string",
        },
        "declaration": {
          "type": "boolean",
        },
        "dir": {
          "type": "string",
        },
        "entryFileNames": {
          "type": "string",
        },
        "externals": {
          "items": {
            "type": "string",
          },
          "type": "array",
        },
        "formats": {
          "items": {
            "enum": [
              "cjs",
              "esm",
              "iife",
              "umd",
              "system",
            ],
            "type": "string",
          },
          "type": "array",
        },
        "globals": {
          "patternProperties": {
            "[a-zA-Z]+": {
              "type": "string",
            },
          },
          "type": "object",
        },
        "meta": {
          "$ref": "#/definitions/meta",
        },
        "sourcemap": {
          "type": "boolean",
        },
      },
      "type": "object",
    },
    "pathRewrite": {
      "items": {
        "additionalProperties": false,
        "properties": {
          "regular": {
            "anyOf": [
              {
                "instanceof": "RegExp",
              },
              {
                "type": "string",
              },
            ],
          },
          "replacement": {
            "type": "string",
          },
        },
        "type": "object",
      },
      "type": "array",
    },
    "proxy": {
      "items": {
        "additionalProperties": true,
        "properties": {
          "changeOrigin": {
            "type": "boolean",
          },
          "pathRewrite": {
            "$ref": "#/definitions/pathRewrite",
          },
          "secure": {
            "type": "boolean",
          },
          "target": {
            "type": "string",
          },
          "url": {
            "type": "string",
          },
        },
        "required": [
          "url",
          "target",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "server": {
      "additionalProperties": false,
      "properties": {
        "host": {
          "type": "string",
        },
        "port": {
          "maximum": 65535,
          "minimum": 0,
          "type": "number",
        },
        "proxy": {
          "$ref": "#/definitions/proxy",
        },
      },
      "type": "object",
    },
  },
  "description": "web-project manifest field in growing-web.",
  "properties": {
    "$schema": {
      "description": "schema remote path.",
      "type": "string",
    },
    "entries": {
      "$ref": "#/definitions/entries",
      "description": "application entries file.",
    },
    "schemaVersion": {
      "description": "web project manifest version.",
      "type": "string",
    },
    "server": {
      "$ref": "#/definitions/server",
      "description": "local server configuration",
    },
  },
  "required": [
    "schemaVersion",
    "entries",
  ],
  "title": "Web Project Manifest",
  "type": "object",
}
`;
